In [3]:
import scipy.stats as stats
import pandas as pd
import seaborn as sns
import pandas as pd
import numpy as np
dataset=sns.load_dataset('tips')

In [ ]:
#t test chi square test paired and unpaired

In [ ]:
normal correlation+ spearman ranked correlation + anova test

# TOC

1. Chi-square test - a. within data b. Goodness of fit
2. T test - a. one-sample b. two-sample c. paired t-test
3. Correlation - a. Pearson b. Spearman
4. Anova Test - a. One way Anova b. Two-way Anova

# Chi square test

## Within the data

In [13]:
from scipy.stats import chi2

In [14]:
dataset.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [15]:
dataset_table=pd.crosstab(dataset['sex'],dataset['smoker'])
print(dataset_table)

smoker  Yes  No
sex            
Male     60  97
Female   33  54


In [16]:
dataset_table.values 

array([[60, 97],
       [33, 54]], dtype=int64)

In [17]:
#Observed Values
Observed_Values = dataset_table.values 
print("Observed Values :-\n",Observed_Values)

Observed Values :-
 [[60 97]
 [33 54]]


In [18]:
val=stats.chi2_contingency(dataset_table)
val

(0.008763290531773594,
 0.925417020494423,
 1,
 array([[59.84016393, 97.15983607],
        [33.15983607, 53.84016393]]))

In [19]:
Expected_Values=val[3]

In [20]:
no_of_rows=len(dataset_table.iloc[0:2,0])
no_of_columns=len(dataset_table.iloc[0,0:2])
ddof=(no_of_rows-1)*(no_of_columns-1)
print("Degree of Freedom:",ddof)
alpha = 0.05

Degree of Freedom: 1


In [21]:
chi_square=sum([(o-e)**2./e for o,e in zip(Observed_Values,Expected_Values)])
chi_square_statistic=chi_square[0]+chi_square[1]

In [22]:
print("chi-square statistic:-",chi_square_statistic)

chi-square statistic:- 0.001934818536627623


In [27]:
critical_value=chi2.ppf(q=1-alpha,df=ddof)
print('critical_value:',critical_value)

critical_value: 3.841458820694124


In [28]:
p_value=1-chi2.cdf(x=chi_square_statistic,df=ddof)
print('p-value:',p_value)
print('Significance level: ',alpha)
print('Degree of Freedom: ',ddof)
print('p-value:',p_value)

p-value: 0.964915107315732
Significance level:  0.05
Degree of Freedom:  1
p-value: 0.964915107315732


In [40]:
if chi_square_statistic>=critical_value:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Fail to Reject H0,There is no relationship between 2 categorical variables")
    
if p_value<=alpha:
    print("Reject H0,There is a relationship between 2 categorical variables")
else:
    print("Fail to Reject H0,There is no relationship between 2 categorical variables")

Fail to Reject H0,There is no relationship between 2 categorical variables
Reject H0,There is a relationship between 2 categorical variables


## Goodness of fit

In [98]:
expected = [50, 50, 50, 50, 50]
observed = [50, 60, 40, 47, 53]

In [100]:
#perform Chi-Square Goodness of Fit Test
statistic, p_value = stats.chisquare(f_obs=observed, f_exp=expected)

In [101]:
if p_value < 0.05:
    print("Reject null hypothesis")
else:
    print("Fail to Reject null hypothesis")

Fail to Reject null hypothesis


In [ ]:
(statistic=4.36, pvaludde=0.35947)

# T Test

## One sample T-test

In [41]:
import numpy as np
from scipy.stats import ttest_1samp

In [42]:
ages=[10,20,35,50,28,40,55,18,16,55,30,25,43,18,30,28,14,24,16,17,32,35,26,27,65,18,43,23,21,20,19,70]

In [43]:
len(ages)

32

In [44]:
ages_mean=np.mean(ages)
print(ages_mean)

30.34375


In [45]:
## Lets take sample

sample_size=10
age_sample=np.random.choice(ages,sample_size)

In [46]:
age_sample

array([25, 18, 16, 70, 10, 28, 23, 18, 27, 30])

In [90]:
ttest,p_value = ttest_1samp(age_sample, 30)

In [91]:
print(p_value)

0.5190525284031675


In [92]:
if p_value < 0.05:
    print("Reject null hypothesis")
else:
    print("Fail to Reject null hypothesis")

Fail to Reject null hypothesis


## Two-sample T-test

In [86]:
np.random.seed(12)
school_ages=stats.poisson.rvs(loc=18,mu=35,size=1500)
classA_height=stats.poisson.rvs(loc=18,mu=30,size=60)

In [87]:
ClassB_ages=stats.poisson.rvs(loc=18,mu=33,size=60)
ClassB_ages.mean()

50.083333333333336

In [88]:
_,p_value = stats.ttest_ind(a=classA_height,b=ClassB_ages,equal_var=False)

In [89]:
if p_value < 0.05:
    print("Reject null hypothesis")
else:
    print("Fail to Reject null hypothesis")

Reject null hypothesis


## Paired T-test

In [ ]:
weight1 = [25,30,28,35,28,34,26,29,30,26,28,32,31,30,45]
weight2 = weight1 + stats.norm.rvs(scale=5,loc=-1.25,size=15)

In [ ]:
print(weight1)
print(weight2)

In [ ]:
weight_df=pd.DataFrame({"weight_10":np.array(weight1),
                         "weight_20":np.array(weight2),
                       "weight_change":np.array(weight2)-np.array(weight1)})

In [ ]:
weight_df

In [ ]:
_,p_value=stats.ttest_rel(a=weight1,b=weight2)

In [ ]:
print(p_value)

In [85]:
if p_value < 0.05:
    print("Reject null hypothesis")
else:
    print("Fail to Reject null hypothesis")

Reject null hypothesis


# Correlation

0.85

In [61]:
df=sns.load_dataset('iris')

In [62]:
df.shape

(150, 5)

## Pearson

In [68]:
df.corr(method = 'pearson')

sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.117570      0.871754     0.817941
sepal_width      -0.117570     1.000000     -0.428440    -0.366126
petal_length      0.871754    -0.428440      1.000000     0.962865
petal_width       0.817941    -0.366126      0.962865     1.000000

## Spearman

In [69]:
df.corr(method = 'spearman')

sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.166778      0.881898     0.834289
sepal_width      -0.166778     1.000000     -0.309635    -0.289032
petal_length      0.881898    -0.309635      1.000000     0.937667
petal_width       0.834289    -0.289032      0.937667     1.000000

# Anova Test (F-Test)

The t-test works well when dealing with two groups, but sometimes we want to compare more than two groups at the same time.

For example, if we wanted to test whether petal_width age differs based on some categorical variable like species, we have to compare the means of each level or group the variable

## One way Anova

In a one-way ANOVA, the one factor or independent variable analyzed has three or more categorical groups

It tell whether two or more groups are similar or not based on their mean similarity and f-score.

Example : There are 3 different category of iris flowers and their petal width and if need to check whether all 3 group are similar or not

In [71]:
df1 = sns.load_dataset('iris')

In [72]:
df1.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [74]:
df_anova = df1[['petal_width','species']]
df_anova.head()

petal_width species
0          0.2  setosa
1          0.2  setosa
2          0.2  setosa
3          0.2  setosa
4          0.2  setosa

In [75]:
grps = pd.unique(df_anova.species.values)
grps

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [78]:
d_data = {grp:df_anova['petal_width'][df_anova.species == grp] for grp in grps}

In [93]:
F, p_value = stats.f_oneway(d_data['setosa'], d_data['versicolor'], d_data['virginica'])

In [96]:
p_value

4.169445839443116e-85

In [97]:
if p_value < 0.05:
    print("Reject null hypothesis")
else:
    print("Fail to Reject null hypothesis")

Reject null hypothesis


## Two way Anova

A two-way ANOVA instead compares multiple groups of two factors

In [106]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [107]:
#create data
df = pd.DataFrame({'water': np.repeat(['daily', 'weekly'], 15),
                   'sun': np.tile(np.repeat(['low', 'med', 'high'], 5), 2),
                   'height': [6, 6, 6, 5, 6, 5, 5, 6, 4, 5,
                              6, 6, 7, 8, 7, 3, 4, 4, 4, 5,
                              4, 4, 4, 4, 4, 5, 6, 6, 7, 8]})
df.head()

water  sun  height
0  daily  low       6
1  daily  low       6
2  daily  low       6
3  daily  low       5
4  daily  low       6

In [108]:
model = ols('height ~ C(water) + C(sun) + C(water):C(sun)', data=df).fit()
sm.stats.anova_lm(model, typ=2)

sum_sq    df        F    PR(>F)
C(water)          8.533333   1.0  16.0000  0.000527
C(sun)           24.866667   2.0  23.3125  0.000002
C(water):C(sun)   2.466667   2.0   2.3125  0.120667
Residual         12.800000  24.0      NaN       NaN

Since the p-values for water and sun are both less than .05, this means that both factors have a statistically significant effect on plant height.

And since the p-value for the interaction effect (.120667) is not less than .05, this tells us that there is no significant interaction effect between sunlight exposure and watering frequency.

# Thank you!